File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    static int answer = 0;
    public int solution(int[] numbers, int target) {
        dfs(0, 0, target, numbers);
        return answer;
    }
    public void dfs(int sum, int cur, int target, int[] numbers) {
        if (cur == numbers.length) {
            if (sum == target) answer++;
            return;
        }
        dfs(sum + numbers[cur], cur + 1, target, numbers);
        dfs(sum - numbers[cur], cur + 1, target, numbers);
        return;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    n = len(numbers)
    def dfs(idx, num):
        
        nonlocal answer
        
        if idx == n - 1:
            if target == num + numbers[idx] or target == num - numbers[idx]:
                answer += 1
            return
        dfs(idx + 1, num + numbers[idx])
        dfs(idx + 1, num - numbers[idx])
    dfs(0, 0)
    return answer